### ISSUE TITLE
Transition the Lexer from using a hardcoded Ruleset to it iterating over an extendable set of Rules

### ISSUE THREAD

**Issue by KilianKilmister** [markedjs/marked#1695]
Currently, when you want to alter the ruleset of the lexer uses, you will have to replace the entirety of the `lexer.blockTokens` and `lexer.inline` functions.

The code used for checking a single rule takes up only a few lines but by the way it's coded, you will need to write out all the unaltered rules again aswell.

Every rule-check looks pretty much alike, for example:

```js
 // fences
 if ((token = this.tokenizer.fences(src))) {
    src = src.substring(token.raw.length)
    tokens.push(token)
    continue
}
```

It would be easy to convert it to iterate over an array of rule-objects and could also be startingpoint for using it asyncronously.

The API would remain the same and performance would stay the same or potentially even improve.

I intend to file a PR for it in the near future, as i need to alter the lexer for my usecase.

### EDIT

As it turnes out, the core if this package is incredibly solid and it's pretty much impossible to modify the parser and lexer from outside the source

**Comment by UziTech** [comment]
If it would have the same performance, or better, I would be ok with it. :+1:

**Comment by KilianKilmister** [comment]
Recently I had to work a lot with large tree structures and last-secon-evaluation and i was able to produce some great results using async generators. They are incredibly pluggable and will seemlessly pass data down the chain. I think they might produce some nice results.

I just forked the repo and i'll see what i can do

**Comment by calculuschild** [comment]
This is one I definitely would like to see for v2.0.0 if it's not any slower. Is the plan to simply place each step of the Lexer into a list and then iterate over the list instead of the pre-determined for-loop we have now? Doesn't sound too hard.... right? Would we want user-provided tokenizers to always be added to the end of the list (maybe right before "text"), or would it make sense to give them an index of where to insert their tokenizer?

@KilianKilmister Did you have any progress with this you would like to contribute to a PR?

**Comment by UziTech** [comment]
It would be nice if the extension would be able to choose where new tokenizers and renderers go in the order. I would like extensions to still be able to just return an object that is sent to `marked.use`. I'm not sure how that would work but maybe something like:

```js
marked.use({
  tokenizer: {
    latex: {
      before: "codespan",
      func: (src) => {
        // latex tokenizer
      },
    }
  }
});
```

**Comment by calculuschild** [comment]
Do we need to support ES5 on this? ES6 allows Maps which iterate in order (as opposed to objects which field order is not guaranteed). Otherwise we would use arrays which just doesn't have the nice benefit of named keys.

**Comment by UziTech** [comment]
I think we will drop ie11 in v2 so we shouldn't need to support es5.

**Comment by UziTech** [comment]
Ya internally we should keep the order in an array.

**Comment by UziTech** [comment]
we need to figure out how to differentiate between inline and block tokenizers/renderers as well

**Comment by calculuschild** [comment]
Another issue is the signature of each item is different. The "code" tokenizer has different arguments (`src, token`) than the "newline" tokenizer (`src`), and I'm not totally sure how to programmatically call the "code" tokenizer with the extra parameter, unless we just give them all the same signature and just pass all parameters to everything?

```
code = function(src, tokens, prevChar, mangle, inRawBlock, etc...) {}
```

That doesn't seem very elegant but I'm not sure of a good way to do it otherwise...

**Comment by calculuschild** [comment]
Or maybe pass in an object with named keys-value pairs? Then the signature of the caller function would stay the same (always just an object) and the individual tokenizers can select which arguments to pass along to tokenizer.js.

```
this.TokenizerParams = {src : src, tokens : tokens, prevChar : prevChar...}

...

code = function(params) {
    if (token = this.tokenizer.code(params.src, params.tokens)) {
      src = src.substring(token.raw.length);
      if (token.type) {
        tokens.push(token);
      } else {
        lastToken = tokens[tokens.length - 1];
        lastToken.raw += '\n' + token.raw;
        lastToken.text += '\n' + token.text;
      }
      return;
    }
}
```

**Comment by UziTech** [comment]
Ya lets see how it affects performance when passing an object with everything to every tokenizer. I also think we should change the parser to send the whole token to the renderer functions instead of having different signatures.

**Comment by UziTech** [comment]
This Should be available in v2.1.0

see Custom Extensions section in [docs](https://marked.js.org/using_pro)
